File: Languages/go/low-level-design-lld/design-patterns/strategy.go
package main

import "fmt"

// Strategy Pattern: Define a family of algorithms (sorting strategies),
// encapsulate each one, and make them interchangeable at runtime.

// Sorter is the strategy interface.
type Sorter interface {
	Sort(data []int) []int
	Name() string
}

// BubbleSort strategy simple O(n²) sort.
type BubbleSort struct{}

func (b BubbleSort) Name() string { return "BubbleSort" }

func (b BubbleSort) Sort(data []int) []int {
	result := make([]int, len(data))
	copy(result, data)
	n := len(result)
	for i := 0; i < n-1; i++ {
		for j := 0; j < n-i-1; j++ {
			if result[j] > result[j+1] {
				result[j], result[j+1] = result[j+1], result[j]
			}
		}
	}
	return result
}

// QuickSort strategy average O(n log n) sort.
type QuickSort struct{}

func (q QuickSort) Name() string { return "QuickSort" }

func (q QuickSort) Sort(data []int) []int {
	result := make([]int, len(data))
	copy(result, data)
	quicksort(result, 0, len(result)-1)
	return result
}

func quicksort(arr []int, low, high int) {
	if low < high {
		pivot := partition(arr, low, high)
		quicksort(arr, low, pivot-1)
		quicksort(arr, pivot+1, high)
	}
}

func partition(arr []int, low, high int) int {
	pivot := arr[high]
	i := low - 1
	for j := low; j < high; j++ {
		if arr[j] <= pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	arr[i+1], arr[high] = arr[high], arr[i+1]
	return i + 1
}

// Context holds a reference to the current strategy.
type SortContext struct {
	strategy Sorter
}

func NewSortContext(s Sorter) *SortContext {
	return &SortContext{strategy: s}
}

func (c *SortContext) SetStrategy(s Sorter) {
	c.strategy = s
}

func (c *SortContext) ExecuteSort(data []int) []int {
	fmt.Printf("Sorting with %s\n", c.strategy.Name())
	return c.strategy.Sort(data)
}

func main() {
	data := []int{64, 34, 25, 12, 22, 11, 90}
	fmt.Printf("Original: %v\n\n", data)

	ctx := NewSortContext(BubbleSort{})
	result := ctx.ExecuteSort(data)
	fmt.Printf("Result:   %v\n\n", result)

	// Swap strategy at runtime
	ctx.SetStrategy(QuickSort{})
	result = ctx.ExecuteSort(data)
	fmt.Printf("Result:   %v\n", result)
}

File: Languages/python/low-level-design-lld/design-patterns/strategy.py
"""
Strategy Pattern

Defines a family of algorithms, encapsulates each one, and makes them
interchangeable. PaymentProcessor can swap strategies at runtime.
"""

from abc import ABC, abstractmethod
from dataclasses import dataclass


class PaymentStrategy(ABC):
    @abstractmethod
    def pay(self, amount: float) -> str:
        pass


class CreditCardPayment(PaymentStrategy):
    def __init__(self, card_number: str):
        self.card_number = card_number

    def pay(self, amount: float) -> str:
        return f"Paid ${amount:.2f} via Credit Card ending {self.card_number[-4:]}"


class PayPalPayment(PaymentStrategy):
    def __init__(self, email: str):
        self.email = email

    def pay(self, amount: float) -> str:
        return f"Paid ${amount:.2f} via PayPal ({self.email})"


class CryptoPayment(PaymentStrategy):
    def __init__(self, wallet: str):
        self.wallet = wallet

    def pay(self, amount: float) -> str:
        return f"Paid ${amount:.2f} via Crypto wallet {self.wallet[:8]}..."


@dataclass
class PaymentProcessor:
    strategy: PaymentStrategy

    def process(self, amount: float) -> str:
        return self.strategy.pay(amount)

    def set_strategy(self, strategy: PaymentStrategy) -> None:
        self.strategy = strategy


if __name__ == "__main__":
    processor = PaymentProcessor(CreditCardPayment("4111222233334444"))
    result = processor.process(99.99)
    assert "Credit Card" in result
    assert "4444" in result

    # Swap strategy at runtime
    processor.set_strategy(PayPalPayment("user@example.com"))
    result = processor.process(50.00)
    assert "PayPal" in result
    assert "user@example.com" in result

    processor.set_strategy(CryptoPayment("0xABCDEF1234567890"))
    result = processor.process(200.00)
    assert "Crypto" in result
    assert "0xABCDEF" in result

    print("All tests passed!")
